sc/src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java

```diff
@@ -71,16 +76,61 @@ public GeneSearchDao(NamedParameterJdbcTemplate namedParameterJdbcTemplate,
         );
     }
 
+    private static final String SELECT_PREFFERED_K_STATEMENT =
+            "SELECT experiment_accession FROM scxa_marker_genes AS markers "+
+            "JOIN scxa_experiment AS experiments ON markers.experiment_accession = experiments.accession "+
+            "WHERE gene_id=:gene_id "+
+            "GROUP BY experiment_accession";
+
+    public Map<String, Map<Integer, List<Integer>>> preferredKAndExperiment (String geneId) {
+        Map<String, Object> namedParameters =
+                ImmutableMap.of(
+                        "gene_id", geneId);
+
+        return namedParameterJdbcTemplate.query(
+                SELECT_PREFFERED_K_STATEMENT,
+                namedParameters,
+                (ResultSet resultSet)->{
+                    Map<String, Map<Integer, List<Integer>>>  result = new HashMap<>();
+
+                    while(resultSet.next()){
+                        String experimentAccession = resultSet.getString("experiment_accession");
+                        IdfParserOutput idfParserOutput = idfParser.parse(experimentAccession);
+                        Integer preferredK = idfParserOutput.getExpectedClusters();
+                        result.putAll(fetchKAndClusterIds(geneId, experimentAccession, preferredK));
+                    }
+
+                    return result;
+
+                }
+        );
+    }
+
+    //In marker gene dataset files, one and only one clusterID for each preferred cluster K value,
+    //which is supposed to have the lowest marker_probability,
+    //so we get one pair from searching (experiment_accesion, k).
+    //If dataset has changed, i.e. there are multiple gene clusterID in one preferred cluster K file,
+    //we probably need to order and limit the searching result based on marker_propability
     private static final String SELECT_K_AND_CLUSTER_ID_FOR_GENE_STATEMENT =
-            "SELECT experiment_accession, k, cluster_id FROM scxa_marker_genes AS markers " +
+            "SELECT experiment_accession, k, marker_probability, cluster_id FROM scxa_marker_genes AS markers " +
```

Comment:
As far as I know, this SQL cannot be simplified anymore, since we have two filter schema and for each accession, we may have one or two return entries.

I have tried what you suggested anyway, but it failed to retrieve the results with two entries.

Reply:
I tried this query straight on the database and it does return 2 rows... 
```
SELECT experiment_accession, k, marker_probability, cluster_id FROM scxa_marker_genes AS markers
WHERE (marker_probability IN (
                               SELECT MIN(marker_probability)
                               FROM scxa_marker_genes
                               WHERE gene_id='ENSG00000013725'
                                 AND experiment_accession='E-GEOD-70580'
                               GROUP BY experiment_accession
                               )
                 AND marker_probability <= 0.05)
     OR ((experiment_accession, k) IN (
      ('E-GEOD-70580', 4))
           AND marker_probability<=0.05
           AND gene_id='ENSG00000013725');
```

In this case, k = 4 is the preferred K value for E-GEOD-70580 (coming from the idf file) and it's not the same as the k for the gene with the minimum p-value.